sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AdaptingIntegerArray.java

```diff
@@ -0,0 +1,221 @@
+        break;
+      case INT:
+        this.intBacking = Arrays.copyOf(toCopy.intBacking, toCopy.intBacking.length);
+        break;
+      case LONG:
+        this.longBacking = Arrays.copyOf(toCopy.longBacking, toCopy.longBacking.length);
+        break;
+    }
+  }
+
+  /**
+   * Copies the size + integer cell size of another array, but allocating a new array of zero
+   * counts.
+   *
+   * @param toZero The other addapting integer array to mirror the size of.
+   */
+  public static AdaptingIntegerArray zeroOf(AdaptingIntegerArray toZero) {
+    return new AdaptingIntegerArray(toZero.length(), toZero.cellSize);
+  }
+
+  /** Add {@code count} to the value stored at {@code index}. */
+  @SuppressWarnings("NullAway")
+  public void increment(int idx, long count) {
+    // TODO - prevent bad index
+    long result;
+    switch (cellSize) {
+      case BYTE:
+        result = byteBacking[idx] + count;
+        if (result > Byte.MAX_VALUE) {
+          // Resize + add
+          resizeToShort();
+          increment(idx, count);
+          return;
+        }
+        byteBacking[idx] = (byte) result;
+        return;
+      case SHORT:
+        result = shortBacking[idx] + count;
+        if (result > Short.MAX_VALUE) {
+          resizeToInt();
+          increment(idx, count);
+          return;
+        }
+        shortBacking[idx] = (short) result;
+        return;
+      case INT:
+        result = intBacking[idx] + count;
+        if (result > Integer.MAX_VALUE) {
+          resizeToLong();
+          increment(idx, count);
+          return;
+        }
+        intBacking[idx] = (int) result;
+        return;
+      case LONG:
+        longBacking[idx] = longBacking[idx] + count;
+        return;
+    }
+  }
+
+  /** Grab the value stored at {@code index}. */
+  @SuppressWarnings("NullAway")
+  public long get(int index) {
+    long value = 0;
+    switch (this.cellSize) {
+      case BYTE:
+        value = this.byteBacking[index];
+        break;
+      case SHORT:
+        value = this.shortBacking[index];
+        break;
+      case INT:
+        value = this.intBacking[index];
+        break;
+      case LONG:
+        value = this.longBacking[index];
+        break;
+    }
+    return value;
+  }
+
+  /** Return the length of this integer array. */
+  @SuppressWarnings("NullAway")
+  public int length() {
+    int length = 0;
+    switch (this.cellSize) {
+      case BYTE:
+        length = this.byteBacking.length;
+        break;
+      case SHORT:
+        length = this.shortBacking.length;
+        break;
+      case INT:
+        length = this.intBacking.length;
+        break;
+      case LONG:
+        length = this.longBacking.length;
+        break;
+    }
+    return length;
+  }
+
+  /** Convert from byte => short backing array. */
+  @SuppressWarnings("NullAway")
+  private void resizeToShort() {
+    this.shortBacking = new short[this.byteBacking.length];
```

Comment:
Let's build up in a local variable and assign to the field

Reply:
Is there a reason, or is this a style thing?

Reply:
Mostly style (setup, then apply), also makes the function more future proof, if locking or volatility changes in the future, it doesn't matter, we see this function acting atomically so easier to reason about it.